<a href="https://colab.research.google.com/github/vfunnny2000/NN/blob/main/HW_8/PD.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Болезнь Паркинсона.

Этот набор данных состоит из ряда биомедицинских измерений голоса людей с болезнью Паркинсона (БП) или без БП. Каждый столбец в таблице представляет собой конкретный показатель голоса, и каждая строка соответствует одной из голосовых записей этих людей (столбец “имя”).

### Вариант 1

In [4]:
import pandas as pd
import numpy as np

In [5]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/parkinsons/parkinsons.data'
data_set = pd.read_csv(url)

In [6]:
print(data_set.shape)
data_set.head()

(195, 24)


name  MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
0  phon_R01_S01_1      119.992       157.302        74.997         0.00784   
1  phon_R01_S01_2      122.400       148.650       113.819         0.00968   
2  phon_R01_S01_3      116.682       131.111       111.555         0.01050   
3  phon_R01_S01_4      116.676       137.871       111.366         0.00997   
4  phon_R01_S01_5      116.014       141.781       110.655         0.01284   

   MDVP:Jitter(Abs)  MDVP:RAP  MDVP:PPQ  Jitter:DDP  MDVP:Shimmer  ...  \
0           0.00007   0.00370   0.00554     0.01109       0.04374  ...   
1           0.00008   0.00465   0.00696     0.01394       0.06134  ...   
2           0.00009   0.00544   0.00781     0.01633       0.05233  ...   
3           0.00009   0.00502   0.00698     0.01505       0.05492  ...   
4           0.00011   0.00655   0.00908     0.01966       0.06425  ...   

   Shimmer:DDA      NHR     HNR  status      RPDE       DFA   spread1  \
0      0.06545  0.02211  21.033       1  0.414783  0.815285 -4.813031   
1      0.09403  0.01929  19.085       1  0.458359  0.819521 -4.075192   
2      0.08270  0.01309  20.651       1  0.429895  0.825288 -4.443179   
3      0.08771  0.01353  20.644       1  0.434969  0.819235 -4.117501   
4      0.10470  0.01767  19.649       1  0.417356  0.823484 -3.747787   

    spread2        D2       PPE  
0  0.266482  2.301442  0.284654  
1  0.335590  2.486855  0.368674  
2  0.311173  2.342259  0.332634  
3  0.334147  2.405554  0.368975  
4  0.234513  2.332180  0.410335  

[5 rows x 24 columns]

In [7]:
data_set.dtypes

name                 object
MDVP:Fo(Hz)         float64
MDVP:Fhi(Hz)        float64
MDVP:Flo(Hz)        float64
MDVP:Jitter(%)      float64
MDVP:Jitter(Abs)    float64
MDVP:RAP            float64
MDVP:PPQ            float64
Jitter:DDP          float64
MDVP:Shimmer        float64
MDVP:Shimmer(dB)    float64
Shimmer:APQ3        float64
Shimmer:APQ5        float64
MDVP:APQ            float64
Shimmer:DDA         float64
NHR                 float64
HNR                 float64
status                int64
RPDE                float64
DFA                 float64
spread1             float64
spread2             float64
D2                  float64
PPE                 float64
dtype: object

Разделяем на данные X и метку y.

In [8]:
X = data_set.iloc[:,1:].drop(labels="status", axis=1)
y = data_set["status"]
print(f'data: {X.shape}\nlabel: {y.shape}')

data: (195, 22)
label: (195,)


Проверка данных на целостность

In [12]:
X.isna().sum(axis=0).sum(axis=0)

0

In [10]:
y.value_counts()
print(f'Количество пациентов с БП: {y.value_counts()[1]}')
print(f'Количество больных БЕЗ БП: {y.value_counts()[0]}')

Количество пациентов с БП: 147
Количество больных БЕЗ БП: 48


In [11]:
X.describe()

MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
count   195.000000    195.000000    195.000000      195.000000   
mean    154.228641    197.104918    116.324631        0.006220   
std      41.390065     91.491548     43.521413        0.004848   
min      88.333000    102.145000     65.476000        0.001680   
25%     117.572000    134.862500     84.291000        0.003460   
50%     148.790000    175.829000    104.315000        0.004940   
75%     182.769000    224.205500    140.018500        0.007365   
max     260.105000    592.030000    239.170000        0.033160   

       MDVP:Jitter(Abs)    MDVP:RAP    MDVP:PPQ  Jitter:DDP  MDVP:Shimmer  \
count        195.000000  195.000000  195.000000  195.000000    195.000000   
mean           0.000044    0.003306    0.003446    0.009920      0.029709   
std            0.000035    0.002968    0.002759    0.008903      0.018857   
min            0.000007    0.000680    0.000920    0.002040      0.009540   
25%            0.000020    0.001660    0.001860    0.004985      0.016505   
50%            0.000030    0.002500    0.002690    0.007490      0.022970   
75%            0.000060    0.003835    0.003955    0.011505      0.037885   
max            0.000260    0.021440    0.019580    0.064330      0.119080   

       MDVP:Shimmer(dB)  ...    MDVP:APQ  Shimmer:DDA         NHR         HNR  \
count        195.000000  ...  195.000000   195.000000  195.000000  195.000000   
mean           0.282251  ...    0.024081     0.046993    0.024847   21.885974   
std            0.194877  ...    0.016947     0.030459    0.040418    4.425764   
min            0.085000  ...    0.007190     0.013640    0.000650    8.441000   
25%            0.148500  ...    0.013080     0.024735    0.005925   19.198000   
50%            0.221000  ...    0.018260     0.038360    0.011660   22.085000   
75%            0.350000  ...    0.029400     0.060795    0.025640   25.075500   
max            1.302000  ...    0.137780     0.169420    0.314820   33.047000   

             RPDE         DFA     spread1     spread2          D2         PPE  
count  195.000000  195.000000  195.000000  195.000000  195.000000  195.000000  
mean     0.498536    0.718099   -5.684397    0.226510    2.381826    0.206552  
std      0.103942    0.055336    1.090208    0.083406    0.382799    0.090119  
min      0.256570    0.574282   -7.964984    0.006274    1.423287    0.044539  
25%      0.421306    0.674758   -6.450096    0.174351    2.099125    0.137451  
50%      0.495954    0.722254   -5.720868    0.218885    2.361532    0.194052  
75%      0.587562    0.761881   -5.046192    0.279234    2.636456    0.252980  
max      0.685151    0.825288   -2.434031    0.450493    3.671155    0.527367  

[8 rows x 22 columns]

In [13]:
import torch
import torch.nn as nn
import torch.optim as optim
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [14]:
# Разделение датасета на тренировочную и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [15]:
# Нормализация данных
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [16]:
# Преобразование данных в тензоры
X_train = torch.tensor(X_train, dtype=torch.float32)
y_train = torch.tensor(y_train.values, dtype=torch.float32)
X_test = torch.tensor(X_test, dtype=torch.float32)
y_test = torch.tensor(y_test.values, dtype=torch.float32)

Создание архитектуры нейросети с двумя линейными слоями и сигмоидной функцией активации.

In [17]:
# Создание нейросети
class ParkNet(nn.Module):
    def __init__(self):
        super(ParkNet, self).__init__()
        self.fc1 = nn.Linear(22, 64)
        self.fc2 = nn.Linear(64, 1)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        x = self.fc1(x)
        x = nn.functional.relu(x)
        x = self.fc2(x)
        x = self.sigmoid(x)
        return x

model = ParkNet()

In [18]:
# Определение функции потерь и оптимизатора
criterion = nn.BCELoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)


In [22]:
# Обучение модели
for epoch in range(100):
    optimizer.zero_grad()
    output = model(X_train)
    loss = criterion(output, y_train.view(-1, 1))
    loss.backward()
    optimizer.step()
    if epoch % 10 == 0:
        print(f'Epoch {epoch}, Loss: {loss.item()}')

Epoch 0, Loss: 0.14048287272453308
Epoch 10, Loss: 0.13524527847766876
Epoch 20, Loss: 0.13011077046394348
Epoch 30, Loss: 0.12511517107486725
Epoch 40, Loss: 0.1202845349907875
Epoch 50, Loss: 0.1155889555811882
Epoch 60, Loss: 0.11109296977519989
Epoch 70, Loss: 0.10675569623708725
Epoch 80, Loss: 0.10253068059682846
Epoch 90, Loss: 0.09843357652425766


In [23]:
# Оценка модели на тестовой выборке
output_test = model(X_test)
predictions = (output_test > 0.5).float()
accuracy = (predictions == y_test.view(-1, 1)).sum().item() / len(y_test)
print(f'Accuracy on test set: {accuracy}')

Accuracy on test set: 0.9487179487179487


Вывод: Точность (Accuracy)  означает, что модель нейросети с двумя линейными слоями и сигмоидной функцией активации правильно классифицирует примерно 94.9% тестовых данных. Высокий уровень точности на тестовом наборе показывает, что модель хорошо обобщает и делает правильные предсказания на новых данных, которые не участвовали в процессе обучения.

## Вариант 2

In [52]:
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Dense


In [53]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/parkinsons/parkinsons.data'
data_set = pd.read_csv(url)

In [54]:
X = data_set.iloc[:,1:].drop(labels="status", axis=1)
y = data_set["status"]

In [55]:
# Создание нейросети
model = Sequential()
model.add(Dense(64, activation='relu', input_dim=X.shape[1]))
model.add(Dense(64, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

In [56]:
# Компилляция модели
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [62]:
# Обучение модели
model.fit(X, y, epochs=10, batch_size=32)

Epoch 1/10
7/7 [==============================] - 0s 3ms/step - loss: 0.3944 - accuracy: 0.8308
Epoch 2/10
7/7 [==============================] - 0s 3ms/step - loss: 0.5222 - accuracy: 0.7590
Epoch 3/10
7/7 [==============================] - 0s 3ms/step - loss: 0.4823 - accuracy: 0.7795
Epoch 4/10
7/7 [==============================] - 0s 3ms/step - loss: 0.5324 - accuracy: 0.8000
Epoch 5/10
7/7 [==============================] - 0s 3ms/step - loss: 0.6903 - accuracy: 0.7231
Epoch 6/10
7/7 [==============================] - 0s 3ms/step - loss: 0.5753 - accuracy: 0.7795
Epoch 7/10
7/7 [==============================] - 0s 3ms/step - loss: 0.4292 - accuracy: 0.8154
Epoch 8/10
7/7 [==============================] - 0s 3ms/step - loss: 0.3845 - accuracy: 0.8410
Epoch 9/10
7/7 [==============================] - 0s 3ms/step - loss: 0.3643 - accuracy: 0.8359
Epoch 10/10
7/7 [==============================] - 0s 3ms/step - loss: 0.3876 - accuracy: 0.8462


In [63]:
# Оценка модели
loss, accuracy = model.evaluate(X, y)
print(f'Loss: {loss}, Accuracy: {accuracy}')

7/7 [==============================] - 0s 4ms/step - loss: 0.5660 - accuracy: 0.7590
Loss: 0.5659653544425964, Accuracy: 0.7589743733406067


In [64]:
print(f'Accuracy on test set: {accuracy}')

Accuracy on test set: 0.7589743733406067


Вывод: модель представляет собой нейронную сеть с двумя скрытыми слоями, каждый из которых содержит 64 нейрона с функцией активации ReLU, и выходным слоем с одним нейроном с функцией активации sigmoid. При обучении данная модель показывает 75.9% точности, но при увеличении количества эпох точность катострофически падает. Модель склонна к переобучению.

***Следовательно вариант № 1 является предпочтительным для решения данной задачи.***



